src/onyx/state/log/bookkeeper.clj

```diff
@@ -0,0 +1,156 @@
+  ([zk-addr zk-root-path timeout]
+   (let [conf (doto (ClientConfiguration.)
+                (.setZkServers zk-addr)
+                (.setZkTimeout timeout)
+                (.setZkLedgersRootPath zk-root-path))]
+     (BookKeeper. conf))))
+
+(def digest-type 
+  (BookKeeper$DigestType/MAC))
+
+(defn password [peer-opts]
+  (.getBytes (arg-or-default :onyx.bookkeeper/ledger-password peer-opts)))
+
+(defrecord BookKeeperLog [client ledger-handle])
+
+;; TODO: add zk-timeout for bookkeeper
+(defmethod state-extensions/initialise-log :bookkeeper [log-type {:keys [onyx.core/replica onyx.core/peer-opts
+                                                                         onyx.core/job-id onyx.core/task-id
+                                                                         onyx.core/kill-ch onyx.core/task-kill-ch
+                                                                         onyx.core/outbox-ch] :as event}] 
+  (let [bk-client (bookkeeper peer-opts)
+        ensemble-size (arg-or-default :onyx.bookkeeper/ledger-ensemble-size peer-opts)
+        quorum-size (arg-or-default :onyx.bookkeeper/ledger-quorum-size peer-opts)
+        ledger-handle (create-ledger bk-client ensemble-size quorum-size digest-type (password peer-opts))
+        slot-id (peer-slot-id event)
+        new-ledger-id (.getId ledger-handle)] 
+    (>!! outbox-ch
+         {:fn :assign-bookkeeper-log-id
+          :args {:job-id job-id
+                 :task-id task-id
+                 :slot-id slot-id
+                 :ledger-id new-ledger-id}})
+    (while (and (first (alts!! [kill-ch task-kill-ch] :default true))
+                (not= new-ledger-id
+                      (last (get-in @replica [:state-logs job-id task-id slot-id]))))
+      (info "New ledger id has not been published yet. Backing off.")
+      (Thread/sleep (arg-or-default :onyx.bookkeeper/ledger-id-written-back-off peer-opts))) 
+    (info "Ledger id published.")
+    (->BookKeeperLog bk-client ledger-handle)))
+
+(defn default-state 
+  "Default state function. Resolves window late for perf."
+  [state window]
+  (if state
+    state
+    ((:window/agg-init window) window)))
+
+(defn playback-windows-extents [state entry windows]
+  (let [id->apply-state-update (into {} 
+                              (map (juxt :window/id :window/apply-state-update) 
+                                   windows))] 
+    (reduce (fn [state' [window-entries {:keys [window/id] :as window}]]
+              (reduce (fn [state'' [extent entry message-id]]
+                        (update-in state'' 
+                                   [id extent]
+                                   (fn [ext-state] 
+                                     (let [ext-state' (default-state ext-state window)
+                                           apply-fn (id->apply-state-update id)] 
+                                       (assert apply-fn (str "Apply fn does not exist for window-id " id))
```

Comment:
I made this an assertion because it should always be true and so it can be turned off. I guess performance isn't totally critical here, so maybe we can make it a non-assertion.